### PR TITLE
[Diffusion] Add Worker contracts and RPC plumbing for Scheduler-managed paged KV cache

### PR DESCRIPTION
**Pull request by Acerak01-fy**
## Purpose

This PR adds the Worker-facing request DTO and RPC plumbing for Scheduler-managed diffusion KV cache. It is the W0 follow-up to merged PR #5541 and the W0 control-plane slice of [RFC #5244](https://github.com/vllm-project/vllm-omni/issues/5244).

The scope is intentionally narrow: this PR defines how a Scheduler allocation snapshot can travel with a newly scheduled diffusion request, while preserving the existing dense execution path. It does not yet create the Scheduler allocator, initialize physical KV pages, install Worker BlockTables, or consume the block IDs in model attention.

## Ownership and data flow

PR #5541 introduced Scheduler-owned logical KV state:

```text
DiffusionKVRequest / DiffusionKVContext
  owner: SchedulerRequestState
  lifetime: mutable across scheduling and native KVCacheManager operations
  rule: never sent to a Worker
```

This PR introduces the corresponding request-time wire snapshot:

```text
DiffusionKVMetadata
  owner: one newly scheduled request envelope
  lifetime: one allocation snapshot
  contents: allocation generation, sequence/context logical bounds,
            native cache-group block IDs
```

The transport path is:

```text
[W1 Scheduler allocation producer]
  -> DiffusionKVMetadata
  -> NewRequestData.diffusion_kv_metadata
  -> MultiprocDiffusionExecutor
       request mode: optional fourth execute_model argument
       batch/step: metadata remains inside DiffusionSchedulerOutput
  -> Diffusion Worker / ModelRunner boundary
  -> [W1 native BlockTable installer]
```

`request.prepared_layout` remains a separate model-owned Worker payload. The generic KV transport does not derive block IDs from it and does not send mutable `DiffusionKVRequest` objects to Workers.

## W0 capability boundary

Metadata is optional in this PR.

Merged PR #5541 stores authoritative `DiffusionKVRequest` objects in `SchedulerRequestState`, removes them from the request sent to the Worker, and does not yet allocate blocks or produce `DiffusionKVMetadata`. Requiring metadata before those producer/installer capabilities exist would make `paged_scheduler` unusable or require fabricated block allocations.

W0 therefore treats these dataclasses as trusted internal DTOs and performs only checks backed by information available at the current execution boundary:

- dense mode rejects unexpected diffusion KV metadata;
- paged mode accepts metadata when supplied but does not require it yet;
- the new-request envelope ID, forwarded request ID, and metadata request ID must agree before model work;
- the historical dense request RPC keeps its existing positional shape.

## Validation ownership

This PR deliberately does not claim strict sequence, cache-group, block-count, or block-range validation. Those checks require independent authoritative state that W0 does not have.

The W1 control-plane and BlockTable follow-up owns these checks at the following boundaries:

```text
Scheduler metadata construction
  -> compare the exact sequence/context set with
     SchedulerRequestState.diffusion_kv_requests
  -> validate sequence IDs, context IDs/roles, and logical lengths

Worker allocation installation
  -> compare cache-group order/count with rank-local native KVCacheConfig
  -> validate required block-table lengths and block ID ranges

Worker allocation registry
  -> enforce allocation_generation idempotency
  -> reject stale or conflicting active snapshots
```

This placement avoids duplicating native `KVCacheConfig`/BlockTable invariants inside a transport DTO and avoids treating DTO-local checks as proof that an allocation matches Scheduler state.

## Main changes

- Add `DiffusionKVContextMetadata`, `DiffusionKVSequenceMetadata`, and  `DiffusionKVMetadata` request allocation DTOs.
- Add optional `diffusion_kv_metadata` to `NewRequestData` so the request and its allocation snapshot cross the Scheduler/Worker boundary atomically.
- Forward request-local metadata through request, request-batch, and step RPC paths without changing the dense request RPC shape.
- Validate request identity against the request object that is actually forwarded.
- Remove the obsolete Stage KV planner digest, physical-layout mirror, Scheduler-owned Worker bootstrap, and initialization acknowledgement types.
- Keep model-owned `prepared_layout` propagation independent from Scheduler allocation metadata.

## Out of scope

- Scheduler `KVCacheManager` allocation, rollback, publication, preemption, or free lifecycle.
- Worker `get_kv_cache_spec()` and available-memory reporting.
- Engine generation and distribution of native `KVCacheConfig` values.
- Physical KV tensor allocation, attention-backend binding, native BlockTables, or slot mappings.
- Sequence/context projection validation and native block geometry/range validation.
- Allocation-generation registry and cleanup lifecycle.
- HunyuanImage3 paged-attention consumption.
- KV connector or cross-stage page transfer.

## Test plan

Focused coverage is under:

```text
tests/diffusion/diffusion_kv/test_metadata.py
tests/diffusion/diffusion_kv/test_worker_contract.py
```

It covers:

- request-local metadata object propagation;
- paged W0 execution without fabricated metadata;
- rejection of dense requests carrying metadata;
- request/envelope/metadata identity mismatches before model work;
- request, batch, and step forwarding behavior;
- preservation of the dense request RPC shape; and
- independent propagation of model-owned `prepared_layout`.

Static checks:

```bash
ruff check \
  vllm_omni/diffusion/diffusion_kv \
  vllm_omni/diffusion/executor/multiproc_executor.py \
  vllm_omni/diffusion/sched/interface.py \
  vllm_omni/diffusion/worker/diffusion_model_runner.py \
  vllm_omni/diffusion/worker/diffusion_worker.py \
  tests/diffusion/diffusion_kv

ruff format --check <same paths>
git diff --check
```

The native cache control plane and BlockTable validation will be covered in W1 against the vLLM version pinned by the target main branch.

### CC List

@zwhzzz0821 @asukaqaq-s @hsliuustc0106

### CODE REVIEW

File: vllm_omni/diffusion/executor/multiproc_executor.py [hunk 1]
```diff
@@ -327,9 +327,13 @@ def execute_request(self, scheduler_output: DiffusionSchedulerOutput) -> BaseRun
         for new_req in scheduler_output.scheduled_new_reqs:
             req = new_req.req
             try:
+                args: tuple = (req, self.od_config, scheduler_output.kv_prefetch_job)
+                stage_kv_metadata = getattr(scheduler_output, "stage_kv_metadata", {}).get(new_req.request_id)
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/executor/multiproc_executor.py]: DiffusionSchedulerOutput now has stage_kv_metadata as a proper field with field(default_factory=dict), so it always exists. The getattr fallback is dead code. Can be simplified to direct access. If the defensive programming against non-DiffusionSchedulerOutput objects is intentional, a comment explaining why would help.

File: vllm_omni/diffusion/stage_kv/interface.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+            raise ValueError("every KV cache group must contain at least one layer")
+        if not isinstance(group.kv_cache_spec, FullAttentionSpec):
+            raise ValueError(
+                "BLOCK_KV_TOKEN_HEAD_DIM_V1 supports only native vLLM "
+                f"FullAttentionSpec, got {type(group.kv_cache_spec).__name__}"
+            )
+        spec = group.kv_cache_spec
+        if spec.block_size <= 0 or spec.num_kv_heads <= 0 or spec.head_size <= 0:
+            raise ValueError(f"invalid FullAttentionSpec geometry for layers {group.layer_names!r}")
+        if spec.head_size_v != spec.head_size:
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/stage_kv/interface.py]: If FullAttentionSpec lacks head_size_v (e.g., a future vLLM API change), this raises AttributeError instead of a clear validation error. Consider getattr(spec, "head_size_v", spec.head_size) for forward compatibility.

File: vllm_omni/diffusion/diffusion_engine.py [hunk 3]
```diff
@@ -220,6 +221,17 @@ def _init_scheduler(
...
+
+    def _initialize_stage_kv_workers(self) -> None:
+        provider = getattr(self.scheduler, "get_stage_kv_worker_init_config", None)
+        if provider is None:
+            return
+        config = provider()
+        if config is None:
+            return
+        result = self.executor.initialize_stage_kv(config)
+        validate_stage_kv_worker_init_result(config, result)
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/diffusion_engine.py]: Both DiffusionExecutor.initialize_stage_kv (abstract.py) and DiffusionEngine._initialize_stage_kv_workers (diffusion_engine.py) call validate_stage_kv_worker_init_result(config, result). The engine-side call is redundant. Consider removing it, or if kept as a defensive boundary, add a comment explaining why.

**Review by RuixiangMa — COMMENTED** [review]
@Acerak01-fy Thanks for your contribution!  I left a few minor nits

File: vllm_omni/diffusion/stage_kv/interface.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+            if not layer_name:
+                raise ValueError("KV cache layer names must be non-empty")
+            if layer_name in specs_by_layer:
+                raise ValueError(f"KV cache layer {layer_name!r} appears in multiple groups")
+            specs_by_layer[layer_name] = spec
+            group_layers.append(layer_name)
+
+    tensor_layers: list[str] = []
+    for tensor in kv_cache_config.kv_cache_tensors:
+        if tensor.size <= 0:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/stage_kv/interface.py]: [P1] Validate the physical tensor geometry

This accepts `KVCacheTensor(size=1, offset=123, block_stride=456)` even though `BLOCK_KV_TOKEN_HEAD_DIM_V1` declares a separate contiguous layout. Upstream allocates from `tensor.size` and treats a positive `block_stride` as packed storage, so W1 can allocate or reshape incompatible memory. Require the exact size implied by `num_blocks * spec.page_size_bytes`, plus the V1 offset/stride semantics, and add negative tests.

File: vllm_omni/diffusion/stage_kv/interface.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+) -> None:
+    """Validate one allocation before it reaches model forward."""
+
+    if not isinstance(config, StageKVWorkerInitConfig):
+        raise TypeError("config must be StageKVWorkerInitConfig")
+    if not isinstance(metadata, StageKVMetadata):
+        raise TypeError("metadata must be StageKVMetadata")
+    if metadata.request_id != request_id:
+        raise ValueError(f"Stage KV metadata request mismatch: expected={request_id!r}, got={metadata.request_id!r}")
+    if metadata.cache_layout_fingerprint != config.cache_layout_fingerprint:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/stage_kv/interface.py]: [P1] Bind metadata to the expected request layout

`request_layout_digest` is currently write-only: #5541 computes it, but this validator has no independently trusted digest to compare against, and `allocation_id` is not tracked either. Stale metadata for a reused request ID can pass when its cache fingerprint and counts match. Pass the expected digest or request generation into this boundary and reject mismatches before model forward.

File: vllm_omni/diffusion/stage_kv/interface.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+    groups = config.kv_cache_config.kv_cache_groups
+    for branch in metadata.branches:
+        if len(branch.block_ids) != len(groups):
+            raise ValueError(
+                f"Stage KV branch {branch.branch_id} has {len(branch.block_ids)} block groups; expected {len(groups)}"
+            )
+        for group_idx, (block_ids, group) in enumerate(zip(branch.block_ids, groups, strict=True)):
+            block_size = group.kv_cache_spec.block_size
+            expected_num_blocks = (branch.resident_len + block_size - 1) // block_size
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/stage_kv/interface.py]: [P1] Validate block count from the planner's `seq_len`

#5541 defines the full first-step `seq_len` as the allocation basis, but this uses `stable_len + current_len`. For block size 4, stable=3, current=1, seq_len=5, the Scheduler correctly allocates 2 blocks and this rejects it as expecting 1. Carry `seq_len` into `StageKVBranchMetadata` and add this boundary case to the combined contract tests.

File: vllm_omni/diffusion/stage_kv/interface.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+            if len(block_ids) != expected_num_blocks:
+                raise ValueError(
+                    f"Stage KV branch {branch.branch_id} group {group_idx} has "
+                    f"{len(block_ids)} blocks; expected {expected_num_blocks}"
+                )
+            if any(block_id < 0 or block_id >= config.kv_cache_config.num_blocks for block_id in block_ids):
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/stage_kv/interface.py]: [P1] Reject vLLM's reserved null block

Upstream `BlockPool` removes block ID 0 from the free queue and reserves it as the null/padding block; worker block tables also use 0 for unused rows. This range check currently accepts 0, so W1 could install a padding block as a real allocation. If this contract reuses upstream `BlockPool`/`BlockTable`, require `0 < block_id < num_blocks` and cover it with a negative test.

**Review by hsliuustc0106 — COMMENTED** [review]
Inline findings from the joint review with #5541.

**Comment by Acerak01-fy** [comment]
> @Acerak01-fy Thanks for your contribution! I left a few minor nits

Thanks! I've fixed it.

**Comment by Acerak01-fy** [comment]
> Inline findings from the joint review with #5541.

Thanks! I have already fixed the issues mentioned.

File: vllm_omni/diffusion/stage_kv/interface.py (hunk 5 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/stage_kv/interface.py]: The expected-digest comparison fixes the write-only part, but the combined contract is still not bound to #5541's retained `StageKVRequirement`. This validator accepts whatever branch set and lengths the metadata self-describes: a focused union probe accepted one metadata branch for a two-branch requirement when given the same expected digest. `allocation_id` is also only checked for positivity, so request-ID reuse is not generation-bound. Please validate against the exact trusted requirement plus a Scheduler-issued allocation/generation identity, with missing-branch and request-reuse negative tests.

File: vllm_omni/diffusion/diffusion_engine.py [hunk 8]
```diff
@@ -220,6 +220,16 @@ def _init_scheduler(
...
+        self._initialize_stage_kv_workers()
+
+    def _initialize_stage_kv_workers(self) -> None:
+        provider = getattr(self.scheduler, "get_stage_kv_worker_init_config", None)
+        if provider is None:
+            return
+        config = provider()
+        if config is None:
+            return
+        self.executor.initialize_stage_kv(config)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/diffusion_engine.py]: [P1] Shut down the executor when Stage KV bootstrap fails

`self.executor` has already launched workers before this call. If config, worker, or acknowledgement validation raises here, `DiffusionEngine.__init__` exits without `close()` or `shutdown()`, while the multiprocess monitor thread keeps the executor alive. This leaves worker processes and GPU allocations behind after failed engine construction. Wrap post-executor initialization in exception cleanup (close an initialized Scheduler and call `executor.shutdown()`), and add a failure-path test.

File: vllm_omni/diffusion/executor/multiproc_executor.py [hunk 9]
```diff
@@ -327,9 +327,19 @@ def execute_request(self, scheduler_output: DiffusionSchedulerOutput) -> BaseRun
         for new_req in scheduler_output.scheduled_new_reqs:
             req = new_req.req
             try:
+                args: tuple = (req, self.od_config, scheduler_output.kv_prefetch_job)
+                stage_kv_metadata = scheduler_output.stage_kv_metadata.get(new_req.request_id)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/executor/multiproc_executor.py]: [P2] Validate the full metadata key set before per-request filtering

These `.get()` calls discard extra request IDs before `DiffusionModelRunner.install_stage_kv_metadata()` can enforce its exact-set check. Batch and step modes forward the full maps, so request mode accepts malformed Scheduler output that they reject. A focused probe with scheduled `req-0` plus `stale` metadata still executed one RPC. Compare both map key sets with the scheduled request IDs before this loop and add an extra-key parity test.

File: vllm_omni/diffusion/executor/abstract.py [hunk 10]
```diff
@@ -89,6 +95,19 @@ def execute_step(self, scheduler_output: DiffusionSchedulerOutput) -> BaseRunner
...
+    def initialize_stage_kv(self, config: StageKVWorkerInitConfig) -> StageKVWorkerInitResult:
+        """Broadcast the Stage KV bootstrap contract to every Worker."""
+
+        validate_stage_kv_worker_init_config(config)
+        responses = self.collective_rpc("initialize_stage_kv", args=(config,))
+        if not isinstance(responses, list) or len(responses) != 1:
+            raise RuntimeError(
+                f"Stage KV initialization expected one rank-0 response envelope, got {type(responses).__name__}"
+            )
+        result = responses[0]
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/executor/abstract.py]: [P2] Validate initialization acknowledgements from every rank

All ranks execute this RPC, but the gathered rank statuses contain only success/error state and this validates only rank 0’s `StageKVWorkerInitResult`. A successful nonzero rank can return a different physical layout, block count, or fingerprint without failing initialization. Gather and validate a compact acknowledgement from every rank, or compare it locally before gathering success, and add a divergent-rank test.

Comment (Acerak01-fy) [reply vllm_omni/diffusion/executor/abstract.py]: Fixed.

File: vllm_omni/diffusion/diffusion_engine.py (hunk 8 above)
Comment (Acerak01-fy) [reply vllm_omni/diffusion/diffusion_engine.py]: fixed.

File: vllm_omni/diffusion/executor/multiproc_executor.py (hunk 9 above)
Comment (Acerak01-fy) [reply vllm_omni/diffusion/executor/multiproc_executor.py]: Fixed.

**Comment by asukaqaq-s** [comment]
What functionality is this PR intended to implement, and where is its scope boundary?

From what I can see, #5550 does not use the following parts of the native vLLM KV-cache initialization and execution flow:

* Worker `get_kv_cache_spec()`
* `determine_available_memory()`
* Engine `get_kv_cache_configs()`
* `generate_scheduler_kv_cache_config()`
* Worker `initialize_from_config()`
* native `BlockTable` / `slot_mapping`
* connector `register_kv_caches()`

Without these pieces, it is not clear to me whether this PR is intended to establish only the control-plane contracts and request-planning abstractions, or whether it is expected to provide a partially functional paged-KV path.

Could you clarify which end-to-end capability this PR is expected to complete, and which parts are intentionally deferred to follow-up PRs?

File: vllm_omni/diffusion/stage_kv/interface.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+            raise ValueError(f"physical layout version must be positive, got {self.version}")
+        if not self.name:
+            raise ValueError("physical layout name must be non-empty")
+
+
+BLOCK_KV_TOKEN_HEAD_DIM_V1 = StageKVPhysicalLayout(
+    version=1,
+    name="BLOCK_KV_TOKEN_HEAD_DIM_V1",
+)
+"""``[num_blocks, 2, block_size, num_kv_heads, head_size]``."""
```
Comment (asukaqaq-s) [inline vllm_omni/diffusion/stage_kv/interface.py]: What does this mean? vLLM FlashAttention uses a logical KV-cache shape of [B, H, N, 2D], while the stride changes depending on whether the physical layout is NHD or HND, as shown in vllm/vllm/v1/attention/backends/flash_attn.py:131. What exactly does physical_layout represent here?

**Comment by asukaqaq-s** [comment]
There seem to be too many `RuntimeError` paths and validation checks in the current implementation, and the same facts are being derived repeatedly across multiple layers.

For example, in request mode, the Executor first performs the metadata/digest exact-set validation in `multiproc_executor.py:327`, and then the Runner performs essentially the same validation again in `diffusion_model_runner.py:194`. The physical configuration is also fully traversed and reconstructed in both the Executor and every Runner, while the resulting values are compared again at both the Worker and Executor layers.

It still makes sense to retain both driver-side preflight validation and receiver-side acceptance validation, especially to avoid a situation where some ranks have already accepted the request before another rank fails. However, both sides should call the same typed envelope and canonical validator, rather than maintaining separate maps and separate missing/unexpected-key loops.

The exception types could also be consolidated according to ownership:

* Use `ValueError` for user configuration or schema errors.
* Use `RuntimeError` for cross-process generation or domain mismatches.
* Preserve the existing assertions or fail-fast behavior for internal invariants owned by native builders and attention backends.

I would suggest retaining only three categories of validation:

1. **Local construction-time invariants**：Non-empty IDs、Positive generation values、Valid enum values、Unique branch IDs

2. **Cross-owner acceptance validation for `StageKVMetadata`**、 Request, domain, and generation matching、Exact branch and group sets、 Block ID ranges、Null-block validation、 Explicit aliasing and unique-writer rules

3. **Real validation during native Worker initialization**、Rank-local layer coverage、Allocation size in bytes、Whether the backend-specific shape, stride, and tensor view can actually be created

**Comment by Acerak01-fy** [comment]
> What functionality is this PR intended to implement, and where is its scope boundary?
> 
> From what I can see, #5550 does not use the following parts of the native vLLM KV-cache initialization and execution flow:
> 
> * Worker `get_kv_cache_spec()`
> * `determine_available_memory()`
> * Engine `get_kv_cache_configs()`
> * `generate_scheduler_kv_cache_config()`
> * Worker `initialize_from_config()`
> * native `BlockTable` / `slot_mapping`
> * connector `register_kv_caches()`
> 
> Without these pieces, it is not clear to me whether this PR is intended to establish only the control-plane contracts and request-planning abstractions, or whether it is expected to provide a partially functional paged-KV path.
> 
> Could you clarify which end-to-end capability this PR is expected to complete, and which parts are intentionally deferred to follow-up PRs?

Thanks, this is a valid concern.

  After discussing this with the author of #5541, we agree that the final initialization ownership should follow the native vLLM flow:

  Worker KV-cache specs and available memory
    -> Engine generates the canonical KVCacheConfig
    -> the same config is consumed by both the Worker physical cache initialization and the Scheduler KVCacheManager.

  Therefore, the current Scheduler-produced get_stage_kv_worker_init_config() flow should not be treated as the final
  architecture. We will align #5541 first and update #5550 before merge so that this PR does not establish a Scheduler-first configuration contract that would immediately need to be replaced.

  The scope of #5550 will remain the control-plane and request-metadata
  boundary:

  - Engine/Executor/Worker initialization plumbing for the shared config;
  - request-local StageKVMetadata and layout-digest delivery;
  - receiver-side validation before model forward;
  - unchanged dense-KV behavior.

  Physical KV tensors, native BlockTables/slot_mapping, model consumption,paged attention, and connector register_kv_caches() will remain in follow-up PRs.

  I will also update the PR description with an explicit scope/ownership table after the shared initialization contract is aligned.

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 12]
```diff
@@ -148,6 +150,24 @@ def __init__(
...
         return getattr(self.pipeline, "device", None)
 
+    def _validate_diffusion_kv_metadata(
+        self,
+        *,
+        request_id: str,
+        metadata: DiffusionKVMetadata | None,
+    ) -> None:
+        cache_mode = getattr(self.od_config, "diffusion_kv_mode", DiffusionKVCacheMode.DENSE_LEGACY)
+        if cache_mode is DiffusionKVCacheMode.PAGED_SCHEDULER:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py]: [P1] This makes the combined paged mode unusable. #5541's `BaseScheduler` still constructs `NewRequestData.from_state(state)` without allocation metadata, so every new request reaches this branch with `None` and fails before request, batch, or step forward. I reproduced that on the merged heads. Keep metadata optional until the Scheduler allocator and Worker installer land behind an explicit initialized capability, then add a Scheduler-to-Runner integration test; placeholder metadata would be wrong because no blocks exist yet.

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 13]
```diff
@@ -584,6 +609,11 @@ def execute_model_batch(
         per-request setup, and calls ``pipeline.forward(batch)``. The pipeline
         must declare ``supports_request_batch = True``.
         """
+        for new_req in scheduler_output.scheduled_new_reqs:
+            self._validate_diffusion_kv_metadata(
+                request_id=new_req.request_id,
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py]: [P2] Validate against the request that is actually forwarded. This compares `metadata.request_id` with `new_req.request_id`, but the batch path then forwards `new_req.req`; the step path has the same split. A focused probe with envelope/metadata ID `envelope-id` and `req.request_id='actual-request-id'` was accepted and forwarded the latter. Require all three IDs to agree before model work, and add batch/step mismatch tests.

File: vllm_omni/diffusion/diffusion_kv/metadata.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    """Worker allocation for an independently managed K/V context."""
+
+    context_id: str
+    cache_role: str
+    num_tokens: int
+    block_ids: tuple[list[int], ...]
+
+
+@dataclass
+class DiffusionKVSequenceMetadata:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/diffusion_kv/metadata.py:31]: [P2] The current PR body promises strict request, group, and block validation, but these DTOs plus `_validate_diffusion_kv_metadata()` enforce only mode and the top-level request ID. A vLLM 0.25 probe accepted negative allocation/sequence/block IDs and `prefix_len + target_len > seq_len`. Validate the logical invariants and exact sequence/context set against #5541's Scheduler-owned requests; before consumption, also validate native group count, block-table lengths, and ID ranges. If this PR is intentionally DTO-only now, update the body and do not require the payload yet.

**Comment by hsliuustc0106** [comment]
The PR description and test plan still describe the old `StageKVWorkerInitConfig/Result`, physical-layout/fingerprint validation, all-Worker initialization RPC, and `install_stage_kv_metadata()` implementation at `9832c66c`; none of those symbols exist at current head `4ab8466a`. Please either restore that advertised initialization contract, or rewrite the title/body/tests for the narrower DTO/RPC-only scope and defer metadata requiredness until the missing producer/installer lands.

**Comment by Acerak01-fy** [comment]
> The PR description and test plan still describe the old `StageKVWorkerInitConfig/Result`, physical-layout/fingerprint validation, all-Worker initialization RPC, and `install_stage_kv_metadata()` implementation at `9832c66c`; none of those symbols exist at current head `4ab8466a`. Please either restore that advertised initialization contract, or rewrite the title/body/tests for the narrower DTO/RPC-only scope and defer metadata requiredness until the missing producer/installer lands.

Done.

File: vllm_omni/diffusion/worker/diffusion_model_runner.py (hunk 12 above)
Comment (Acerak01-fy) [reply vllm_omni/diffusion/worker/diffusion_model_runner.py]: fixed.

File: vllm_omni/diffusion/worker/diffusion_model_runner.py (hunk 13 above)
Comment (Acerak01-fy) [reply vllm_omni/diffusion/worker/diffusion_model_runner.py]: fixed.

File: vllm_omni/diffusion/diffusion_kv/metadata.py (hunk 14 above)
Comment (Acerak01-fy) [reply vllm_omni/diffusion/diffusion_kv/metadata.py:31]: fixed.

File: vllm_omni/diffusion/diffusion_kv/metadata.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    sequence_id: int
+    prefix_len: int
+    target_len: int
+    seq_len: int
+    block_ids: tuple[list[int], ...]
+    contexts: tuple[DiffusionKVContextMetadata, ...] = ()
+
+
+@dataclass
+class DiffusionKVMetadata:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/diffusion_kv/metadata.py:47]: do we have seperate arg for CFGs?

Comment (Acerak01-fy) [reply vllm_omni/diffusion/diffusion_kv/metadata.py:47]: The current design has already taken CFGs into account.

**Comment by hsliuustc0106** [comment]
```

[2026-08-10T14:14:45Z] FAILED tests/diffusion/cache/test_teacache_extractors.py::TestMiniMaxH3Extractor::test_modulated_input_shape - AttributeError: type object 'FakeBackend' has no attribute 'supports_packed_mask_free'
--
  | [2026-08-10T14:14:45Z] FAILED tests/diffusion/cache/test_teacache_extractors.py::TestMiniMaxH3Extractor::test_run_transformer_blocks_callable - AttributeError: type object 'FakeBackend' has no attribute 'supports_packed_mask_free'
  | [2026-08-10T14:14:45Z] FAILED tests/diffusion/cache/test_teacache_extractors.py::TestMiniMaxH3Extractor::test_postprocess_callable - AttributeError: type object 'FakeBackend' has no attribute 'supports_packed_mask_free'
  | [2026-08-10T14:14:45Z] FAILED tests/diffusion/cache/test_teacache_extractors.py::TestMiniMaxH3Extractor::test_postprocess_output_shape - AttributeError: type object 'FakeBackend' has no attribute 'supports_packed_mask_free'
  | [2026-08-10T14:14:45Z] FAILED tests/diffusion/cache/test_teacache_extractors.py::TestMiniMaxH3Extractor::test_extractor_matches_native_forward - AttributeError: type object 'FakeBackend' has no attribute 'supports_packed_mask_free'
  | [2026-08-10T14:14:45Z] FAILED tests/diffusion/cache/test_teacache_extractors.py::TestMiniMaxH3Extractor::test_run_transformer_blocks_forwards_packed_layout - AttributeError: type object 'FakeBackend' has no attribute 'supports_packed_mask_free'
  | [2026-08-10T14:14:45Z] FAILED tests/diffusion/cache/test_teacache_extractors.py::TestMiniMaxH3Extractor::test_teacache_reuses_h3_block_residual - AttributeError: type object 'FakeBackend' has no attribute 'supports_packed_mask_free'
  | [2026-08-10T14:14:45Z] = 7 failed, 2710 passed, 30 skipped, 366 deselected, 31 warnings in 463.86s (0:07:43) =
  | [2026-08-10T14:14:49Z] /usr/lib/python3.12/multiprocessing/resource_tracker.py:279: UserWarning: resource_tracker: There appear to be 1 leaked shared_memory objects to clean up at shutdown
  | [2026-08-10T14:14:49Z]   warnings.warn('resource_tracker: There appear to be %d '
  | [2026-08-10T14:14:52Z] 🚨 Error: The command exited with status 1
  | [2026-08-10T14:14:52Z] user command error: The plugin docker command hook exited with status 1


```
